### ISSUE TITLE
Provide direct access to all tokens: id_token, access_token and refresh_token

### ISSUE THREAD

**Issue by apulliam** [aspnet/Security#1784]
The OpenID Connect middleware can be configured with any OpenIdConnectResponseType, i.e.:

   options.ResponseType = OpenIdConnectResponseType.IdTokenToken

However, using the above setting, the middleware doesn't appear give any way to get the access_token.  For the hyrid auth code flow, all the docs show adding a handler to manually exchange the auth code for the access_token (using ADAL or MSAL):
     OnAuthorizationCodeReceived = async (context) =>
                    {
                        var code = context.ProtocolMessage.Code;
                        var identifier = context.Principal.FindFirst(Startup.ObjectIdentifierType).Value;
                        var memoryCache = context.HttpContext.RequestServices.GetRequiredService<IMemoryCache>();
                        var graphScopes = _azureOptions.GraphScopes.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                        
                        var cca = new ConfidentialClientApplication(
                            _azureOptions.Aadv2ClientId, 
                            _azureOptions.BaseUrl + _azureOptions.CallbackPath,
                            new ClientCredential(_azureOptions.Aadv2ClientSecret),
                            new AzureAd2SessionTokenCache(identifier, memoryCache).GetCacheInstance(), 
                            null);
                        var result = await cca.AcquireTokenByAuthorizationCodeAsync(code, graphScopes);
                      
                        context.HandleCodeRedemption(result.AccessToken, result.IdToken);
                    },

I'd like to be able to let the OpenID connect middleware complete the hybrid implicit or auth code flow and get access to the access_token, refresh_token (for auth code).  

I don't understand why allow the middleware is able to complete the implicit or auth code flows automatically, but doesn't provide access to the tokens.  

This doesn't make any sense to me.  Having to write code to manually exchange the auth code for the access_token seems messy and error prone.  I see this provides access to a token cache, but the middleware could do this as well.

**Comment by Tratcher** [comment]
Provide access to them where? SaveTokens?

**Comment by apulliam** [comment]
The purpose SaveTokens isn't clear.  The comment is "Defines whether access and refresh tokens should be stored in the Microsoft.AspNetCore.Http.Authentication.AuthenticationProperties after a successful authorization."  How can I get access to AuthenticationProperties in order to get access to the saved tokens?  As to where in the code I would like to access the tokens, I would like to access them in the appropriate event handler.  OnTicketReceived seems to be the last one called.

**Comment by Tratcher** [comment]
In OnTicketReceived, the context object has the AuthenticationProperties on it. You can get tokens from there by calling GetToken(name).
This sample does something similar:
https://github.com/aspnet/AuthSamples/blob/404105bd191f2e973d4befb668f1310d2fd82701/samples/Identity.ExternalClaims/Startup.cs#L48

**Comment by apulliam** [comment]
Thanks! This is really obscure :-), but exactly what I was looking for.  There are situations where letting the OpenID Connect middleware complete the OAuth flow makes more sense than including ADAL and MSAL and splitting the flow between 2 libraries.  I just needed to know how to get to the tokens!

**Comment by apulliam** [comment]
Is there also a way to access AuthenticationProperties from a Controller?  If not, what is the recommended way to pass state from middelware (OnTicketReceived) to the Controller?

**Comment by Tratcher** [comment]
Yes, you call HttpContext.AuthenticateAsync().

**Comment by apulliam** [comment]
Cool. Again, very useful but needs documentation. By the name it sounds like its possibly doing another authentication (if necessary).  Does this work like ADAL/MSAL and refresh the access token in the background?

**Comment by Tratcher** [comment]
No, it's returning cached results from prior calls to AuthenticateAsync on that request.

**Comment by apulliam** [comment]
I would suggest a different name for the method then, perhaps GetAuthenticationProperties with or without the Async suffix.
